Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
+      spectrum.setDetectorIDs(
+          std::set<detid_t>(detectorIDs[i].begin(), detectorIDs[i].end()));
+  }
+}
```

Comment:
Keeping it will be a second step. If we would keep it right now, we would have two locations where spectrum numbers and detector IDs are stored. As long as `ISpectrum::setSpectrumNo` and `ISpectrum::setDetectorIDs` are still available we cannot keep the translator, since the information in the translator will run out of sync with the information in the spectra.